src/script/lua_api/l_object.cpp

```diff
@@ -252,6 +252,11 @@ int ObjectRef::l_set_hp(lua_State *L)
 		reason.lua_reference = luaL_ref(L, LUA_REGISTRYINDEX);
 	}
 
+	// Abort operation if attempting to set HP higher than hp_max
+	int hp_max = co->accessObjectProperties()->hp_max;
```

Comment:
Please extend `PlayerSAO::setHP` with this check, preferably before this Lua callback is executed:
https://github.com/minetest/minetest/blob/ae1caba6aaa6f0a13e70f195f371fd4e7a378491/src/content_sao.cpp#L1325

Reply:
Dunno, I thought it's easier to implement in l_object.cpp

Is there a reason to add this to PlayerSAO::setup instead? :) 

Also note that this will then only apply to players, and entities would be able to bypass *_max properties.

Reply:
@ClobberXD That's the idea of it. `hp_max` is only for players. For entities the mods control the maximal health (if it ever regains).

Reply:
`hp_max` is a part of object properties; shouldn't it apply to all ObjectRefs?

Reply:
Question is whether it would make sense. Mobs don't gain health by the engine - their HP is entirely controlled by mods, so I don't really see a practical use for that.

Reply:
Fair point... will move the checks to content_sao.cpp

Reply:
Done